common/src/main/java/bisq/common/crypto/KeyStorage.java

```diff
@@ -150,20 +209,49 @@ public KeyPair loadKeyPair(KeyEntry keyEntry) {
         }
     }
 
-    public void saveKeyRing(KeyRing keyRing) {
-        savePrivateKey(keyRing.getSignatureKeyPair().getPrivate(), KeyEntry.MSG_SIGNATURE.getFileName());
-        savePrivateKey(keyRing.getEncryptionKeyPair().getPrivate(), KeyEntry.MSG_ENCRYPTION.getFileName());
+    public void saveKeyRing(KeyRing keyRing, String password) {
+        savePrivateKey(keyRing.getSignatureKeyPair().getPrivate(), KeyEntry.MSG_SIGNATURE.getFileName(), password);
+        savePrivateKey(keyRing.getEncryptionKeyPair().getPrivate(), KeyEntry.MSG_ENCRYPTION.getFileName(), password);
     }
 
-    private void savePrivateKey(PrivateKey privateKey, String name) {
+    private void savePrivateKey(PrivateKey privateKey, String name, String password) {
         if (!storageDir.exists())
             //noinspection ResultOfMethodCallIgnored
-            storageDir.mkdir();
+            storageDir.mkdirs();
 
         PKCS8EncodedKeySpec pkcs8EncodedKeySpec = new PKCS8EncodedKeySpec(privateKey.getEncoded());
         try (FileOutputStream fos = new FileOutputStream(storageDir + "/" + name + ".key")) {
-            fos.write(pkcs8EncodedKeySpec.getEncoded());
-        } catch (IOException e) {
+            byte[] keyBytes = pkcs8EncodedKeySpec.getEncoded();
+            // Encrypt
+            if (password != null) {
+                // Magic
+                fos.write(ENCRYPTED_FORMAT_MAGIC);
+
+                // Version, salt length, iterations, and secret length.
+                ByteBuffer header = ByteBuffer.allocate(ENCRYPTED_FORMAT_LENGTH);
+                header.putInt(ENCRYPTED_FORMAT_VERSION);
+                header.putInt(SALT_LENGTH);
+                header.putInt(SECRET_ITERATIONS);
+                header.putInt(SECRET_LENGTH);
+                fos.write(header.array());
+
+                // Write pw salt and pw encrypted
+                byte[] salt = CryptoUtils.getRandomBytes(SALT_LENGTH);
+                fos.write(salt);
+                KeyCrypterScrypt crypter = ScryptUtil.getKeyCrypterScrypt(salt);
+                KeyParameter pwKey = ScryptUtil.deriveKeyWithScrypt(crypter, password);
+                fos.write(pwKey.getKey());
+
+                // Write new salt and generate SecretKey
+                salt = CryptoUtils.getRandomBytes(SALT_LENGTH);
+                fos.write(salt);
+                SecretKey secretKey = Encryption.generateSecretKey(password, salt, SECRET_ITERATIONS, SECRET_LENGTH);
```

Comment:
`Encryption.generateSecretKey`  and `ScryptUtil.deriveKeyWithScrypt` seems to be doing something very similar - using the same password, different salt and different algorithm, it derives a key. Is it really necessary to use both? I'm probably missing something

Reply:
The Scrypt algorithm is important when persisting the password hash because it is much slower than standard key derivation and also requires memory. Since I persist that Scrypt derived key to validate the password, it must be different from the key used to encrypt the user's private keys (which is used to identify a particular user in the p2p system). 

If I persist the same derived key, the encryption of the user's private keys would be pointless as an attacker with the files will immediately decrypt the private key data and be able to impersonate the user.

Reply:
Unified encryption key with password checking.